pilot/pkg/proxy/envoy/v2/ads_common.go

```diff
@@ -0,0 +1,181 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/schemas"
+)
+
+func ProxyNeedsPush(proxy *model.Proxy, pushEv *XdsEvent) bool {
+	if !features.ScopePushes.Get() {
+		// If push scoping is not enabled, we push for all proxies
+		return true
+	}
+
+	targetNamespaces := pushEv.namespacesUpdated
+	configs := pushEv.configTypesUpdated
+
+	// appliesToProxy starts as false, we will set it to true if we encounter any configs that require a push
+	appliesToProxy := false
+	// If no config specified, this request applies to all proxies
+	if len(configs) == 0 {
+		appliesToProxy = true
+	}
+Loop:
+	for config := range configs {
+		switch config {
+		case schemas.Gateway.Type:
+			if proxy.Type == model.Router {
+				return true
+			}
+		case schemas.QuotaSpec.Type, schemas.QuotaSpecBinding.Type:
+			if proxy.Type == model.SidecarProxy {
+				return true
+			}
+		default:
+			appliesToProxy = true
+			break Loop
+		}
+	}
+
+	if !appliesToProxy {
+		return false
+	}
+
+	// If no only namespaces specified, this request applies to all proxies
+	if len(targetNamespaces) == 0 {
+		return true
+	}
+
+	// Otherwise, only apply if the egress listener will import the config present in the update
+	for ns := range targetNamespaces {
+		if proxy.SidecarScope.DependsOnNamespace(ns) {
+			return true
+		}
+	}
+	return false
+}
+
+type XdsType int
+
+const (
+	CDS XdsType = iota
+	EDS
+	LDS
+	RDS
+)
+
+// TODO: merge with ProxyNeedsPush
+func PushTypeFor(proxy *model.Proxy, pushEv *XdsEvent) map[XdsType]bool {
+	out := map[XdsType]bool{}
+
+	// In case configTypes is not set, for example mesh configuration updated.
+	// If push scoping is not enabled, we push all xds
+	if !features.ScopePushes.Get() || len(pushEv.configTypesUpdated) == 0 {
+		out[CDS] = true
+		out[EDS] = true
+		out[LDS] = true
+		out[RDS] = true
+		return out
+	}
+
+	if proxy.Type == model.SidecarProxy {
+		for config := range pushEv.configTypesUpdated {
+			switch config {
+			case schemas.VirtualService.Type:
+				out[LDS] = true
+				out[RDS] = true
+			case schemas.Gateway.Type:
+				// Do not push
+			case schemas.ServiceEntry.Type, schemas.SyntheticServiceEntry.Type:
+				out[CDS] = true
+				out[EDS] = true
+				out[LDS] = true
+				out[RDS] = true
+			case schemas.DestinationRule.Type:
+				out[CDS] = true
+				out[EDS] = true
+			case schemas.EnvoyFilter.Type:
+				out[CDS] = true
+				out[LDS] = true
+				out[RDS] = true
+			case schemas.Sidecar.Type:
```

Comment:
need eds

Reply:
same as above

Reply:
I still think there are some cases... let me take a look at the code and see if I can find one

Reply:
Maybe the https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_istio/18354/pilot-e2e-envoyv2-v1alpha3_istio/3054/artifacts/kind/istio-testing-control-plane/pods/istio-system_b-56885994db-5zqqn_e0ea7570-3fa7-43e8-be52-3f55f95c4dcf/istio-proxy/0.log  `UH` error maybe related to eds. Will let ci test it.

Reply:
But in my local env, i can not reproduce the `pilot-e2e-envoyv2-v1alpha3_istio` failure.